### DISCUSSION TITLE
How to delete temporary file with upload file after write to other path

### DISCUSSION THREAD

**Discussion by arimakai** [fastapi/fastapi#10607] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@router.post("/test/upload")
async def post_endpoint(testfile: UploadFile):
    buffer = await testfile.read()
    filesize = len(buffer)
    async with aiofiles.open('/storage1/testfile', 'wb') as output_file:
        await output_file.write(buffer)  # async write

    return {"filename": testfile.filename, "filesize": filesize}
```

### Description

Before run program

![image](https://github.com/tiangolo/fastapi/assets/54058054/5145cd3c-f2af-4ee4-aff7-d97f59fe80f3)

After upload file

![image](https://github.com/tiangolo/fastapi/assets/54058054/76d8020e-68c5-4f84-b843-7b4520040bdc)

My disk space will be used up either when uploading a new file or when canceling the program.
It is possible to delete immediately before responding.

### Operating System

Linux

### Operating System Details

Debian 11

### FastAPI Version

0.77.1

### Pydantic Version

1.9.0

### Python Version

3.9.2

**Comment by jgould22** [comment]
You should use https://docs.python.org/3/library/tempfile.html

aiofiles provides an interface for the above to use in an async context

Though in your example code the `testfile` will already be in temp storage

**Comment by arimakai** [comment]
Dear jgould22,

Thank you for your reply. Can you suggest if I change to the tempfile library, whether the other API paths will stop with synchronous pattern during file writes or not?

Regards,
Arima

**Reply by jgould22** [reply]
Do not use tempfile directly.

You can use aiofiles like you do in your example https://github.com/Tinche/aiofiles#tempfile

**Reply by arimakai** [reply]
Thank you for your suggestion. 
I still have a problem when the upload URL path is run. 
I cannot directly save temporary files to other paths outside the root "/" partition. 
I think the default behavior of Python or FastAPI is to save temporary files to the root partition first, and then we can move or copy them to another directory. 
After I close the file with await testfile.close() (from my code above), I will regain the free space occupied by the temporary file. However, this means that the temporary file is using the root partition's storage space.

**Reply by arimakai** [reply]
OK,
I found the solution to change the temporary file from this link
https://stackoverflow.com/questions/71374765/how-to-select-the-disk-location-for-uploadfile-parameter-in-fastapi

Regards,
Arima
